File: leetcode/easy/DesignHashMap.java
class MyHashMap {

    // --- Inner Node Class for Linked List ---
    // Represents a key-value pair in the hash map.
    // Made private static to encapsulate and avoid implicit reference to MyHashMap instance.
    private static class Node {
        final int key;
        int value;
        Node next; // Pointer to the next node in case of collision

        public Node(int key, int value) {
            this.key = key;
            this.value = value;
            this.next = null;
        }
    }

    // --- MyHashMap Fields ---
    private Node[] buckets; // Array of Node heads (each Node is the head of a linked list)
    private int capacity;          // Current capacity of the hash table (number of buckets)
    private int size;              // Current number of key-value mappings in the hash map
    private final double LOAD_FACTOR_THRESHOLD_UP = 0.75; // Threshold for resizing up
    private final double LOAD_FACTOR_THRESHOLD_DOWN = 0.25; // Threshold for resizing down
    private final int INITIAL_CAPACITY = 16; // Initial capacity, often a power of 2

    /**
     * Initializes a new instance of MyHashMap.
     */
    public MyHashMap() {
        this.capacity = INITIAL_CAPACITY;
        this.buckets = new Node[this.capacity]; // Initialize array of Node heads
        // Note: Each element in `buckets` is initially null. Nodes are added on `put`.
        this.size = 0;
    }

    /**
     * Computes the hash index for a given key.
     * Uses a simple modulo operation. Assumes non-negative keys for simplicity.
     * For negative keys, a more robust hash would be `(Integer.hashCode(key) & 0x7fffffff) % this.capacity`.
     *
     * @param key The key to hash.
     * @return The index in the buckets array.
     */
    private int hash(int key) {
        return key % this.capacity;
    }

    /**
     * Inserts a key-value pair into the hash map. If the key already exists,
     * its value is updated. Otherwise, a new entry is added.
     *
     * @param key The key to insert or update.
     * @param value The value associated with the key.
     */
    public void put(int key, int value) {
        int index = hash(key);
        Node current = this.buckets[index]; // Head of the linked list at this bucket

        // Traverse the linked list to find if the key already exists.
        while (current != null) {
            if (current.key == key) {
                current.value = value; // Key found, update its value.
                return;
            }
            current = current.next;
        }

        // Key does not exist, add a new node at the beginning of the bucket's linked list.
        Node newNode = new Node(key, value);
        newNode.next = this.buckets[index];
        this.buckets[index] = newNode;
        this.size++;

        // Check if resizing (expanding) is needed after adding a new entry.
        if ((double) this.size / this.capacity > LOAD_FACTOR_THRESHOLD_UP) {
            resize(this.capacity * 2);
        }
    }

    /**
     * Retrieves the value associated with the given key.
     *
     * @param key The key to search for.
     * @return The value associated with the key, or -1 if the key is not found.
     */
    public int get(int key) {
        int index = hash(key);
        Node current = this.buckets[index]; // Head of the linked list at this bucket

        // Traverse the linked list to find the key.
        while (current != null) {
            if (current.key == key) {
                return current.value; // Key found, return its value.
            }
            current = current.next;
        }
        return -1; // Key not found.
    }

    /**
     * Removes the key-value pair associated with the given key from the hash map.
     *
     * @param key The key to remove.
     */
    public void remove(int key) {
        int index = hash(key);
        Node head = this.buckets[index]; // Head of the linked list at this bucket

        // Use a dummy node to simplify removal, especially if the head of the list needs to be removed.
        Node dummy = new Node(-1, -1); // Dummy node, key/value don't matter
        dummy.next = head;

        Node prev = dummy;
        Node current = head;

        // Traverse the linked list to find the node to remove.
        while (current != null && current.key != key) {
            prev = current;
            current = current.next;
        }

        // If current is not null, it means we found the node to remove.
        if (current != null) {
            prev.next = current.next; // Bypass the node to remove.
            this.buckets[index] = dummy.next; // Update the bucket head (in case head was removed).
            this.size--;

            // Check if resizing (shrinking) is needed after removing an entry.
            // Only shrink if capacity is not already at its minimum (INITIAL_CAPACITY).
            if (this.capacity > INITIAL_CAPACITY && (double) this.size / this.capacity < LOAD_FACTOR_THRESHOLD_DOWN) {
                resize(this.capacity / 2);
            }
        }
    }

    /**
     * Resizes the hash table to a new capacity and rehashes all existing entries.
     *
     * @param newCapacity The new capacity for the hash table.
     */
    private void resize(int newCapacity) {
        // Store old buckets temporarily.
        Node[] oldBuckets = this.buckets;
        int oldCapacity = this.capacity;

        // Update capacity and initialize new buckets array.
        this.capacity = newCapacity;
        this.buckets = new Node[this.capacity];
        // No need to explicitly initialize each bucket to null, as it's the default for object arrays.
        
        this.size = 0; // Reset size, as put() will increment it again during rehash.

        // Rehash all entries from old buckets into new buckets.
        for (int i = 0; i < oldCapacity; i++) {
            Node node = oldBuckets[i];
            while (node != null) {
                this.put(node.key, node.value); // Use put method to rehash each entry
                node = node.next;
            }
        }
    }
}

File: leetcode/easy/DesignHashMap.py
from typing import Optional


class _Node:
    """A node in a singly linked list, used for chaining in the HashMap.
    
    Each node stores a key-value pair and a pointer to the next node in the
    same bucket, handling hash collisions.
    """
    def __init__(self, key: int, value: int, next_node: Optional['_Node'] = None):
        self.key = key
        self.value = value
        self.next = next_node


class MyHashMap:
    """A custom implementation of a HashMap using separate chaining for collision resolution.

    This HashMap supports 'put', 'get', and 'remove' operations. It also dynamically
    resizes itself by doubling or halving its capacity based on a load factor to
    maintain average O(1) time complexity for its operations.
    """

    def __init__(self):
        """Initializes the HashMap with a default capacity and empty buckets."""
        self._INITIAL_CAPACITY = 16
        self._LOAD_FACTOR_THRESHOLD_UP = 0.75
        self._LOAD_FACTOR_THRESHOLD_DOWN = 0.25
        
        self.capacity = self._INITIAL_CAPACITY
        self.size = 0
        # Each bucket will be the head of a linked list of _Node objects.
        self.buckets = [None] * self.capacity

    def _hash(self, key: int) -> int:
        """Computes the bucket index for a given key using a simple modulo operation."""
        return key % self.capacity

    def put(self, key: int, value: int) -> None:
        """Inserts a key-value pair. Updates the value if the key already exists.
        
        It traverses the linked list at the corresponding bucket. If the key is found,
        it updates the value. Otherwise, it adds a new node to the head of the list.
        After insertion, it checks if a resize is needed.
        """
        index = self._hash(key)
        curr = self.buckets[index]

        # Traverse the list to find if the key already exists.
        while curr:
            if curr.key == key:
                curr.value = value  # Key found, update value and return.
                return
            curr = curr.next

        # Key not found, add a new node to the beginning of the bucket's list.
        new_node = _Node(key, value, self.buckets[index])
        self.buckets[index] = new_node
        self.size += 1

        # Check if resizing (expanding) is needed.
        if (self.size / self.capacity) > self._LOAD_FACTOR_THRESHOLD_UP:
            self._resize(self.capacity * 2)

    def get(self, key: int) -> int:
        """Retrieves the value for a key.
        
        Returns:
            The value associated with the key, or -1 if the key is not found.
        """
        index = self._hash(key)
        curr = self.buckets[index]

        # Traverse the list to find the key.
        while curr:
            if curr.key == key:
                return curr.value  # Key found, return its value.
            curr = curr.next
        
        return -1  # Key not found.

    def remove(self, key: int) -> None:
        """Removes the key-value pair for the given key if it exists."""
        index = self._hash(key)
        
        # Use a dummy node to simplify removal, especially for the head of the list.
        dummy = _Node(-1, -1, self.buckets[index])
        prev = dummy
        curr = self.buckets[index]

        # Traverse the list to find the node to remove.
        while curr:
            if curr.key == key:
                # Key found, bypass the current node to remove it.
                prev.next = curr.next
                self.buckets[index] = dummy.next  # Update the bucket head.
                self.size -= 1

                # Check if resizing (shrinking) is needed.
                if self.capacity > self._INITIAL_CAPACITY and (self.size / self.capacity) < self._LOAD_FACTOR_THRESHOLD_DOWN:
                    self._resize(self.capacity // 2)
                return
            
            prev = curr
            curr = curr.next

    def _resize(self, new_capacity: int) -> None:
        """Resizes the hash table and re-hashes all existing entries."""
        old_buckets = self.buckets

        # Reset properties for the new table structure.
        self.capacity = new_capacity
        self.buckets = [None] * self.capacity
        self.size = 0

        # Re-hash all entries from the old buckets into the new buckets.
        for bucket_head in old_buckets:
            curr = bucket_head
            while curr:
                self.put(curr.key, curr.value)
                curr = curr.next